### ISSUE TITLE
Assignments with operators duplicate code/calls

### ISSUE THREAD

**Issue by demipixel** [godotengine/godot#21832]
**Godot version:**

3.0.6

**OS/device including version:**
Windows 10, should have no impact

**Issue description:**
Assignments with operators duplicate code (or at least calls). For example, `arr[randi() % n] += 1` calls `randi()` twice (you can use a custom function to test this), presumably from doing `arr[randi() % n] = arr[randi() % n] + 1`.

Basic example:
```
extends Node

func _ready():
	var arr = [0, 1]
	arr[get_num()] += 2
	
	print(arr)
	# Expected: [2, 1]
	# Actual:   [3, 1]

var index_to_return = 0
func get_num():
	index_to_return += 1
	return index_to_return - 1
```

**Comment by demipixel** [comment]
Alright, I may be wrong here, but from what I can tell, we parse the destination address here:

https://github.com/godotengine/godot/blob/f06b7d40c87e6a2435cdfcad6bb620184f16ea42/modules/gdscript/gdscript_compiler.cpp#L1139

Then a few lines down call `_parse_assign_right_expression()` which calls `_parse_expression()` which will again parse the expression:

https://github.com/godotengine/godot/blob/f06b7d40c87e6a2435cdfcad6bb620184f16ea42/modules/gdscript/gdscript_compiler.cpp#L97

**Comment by aaronfranke** [comment]
The source and desetination being the same should only the case for `thing += x` operators but I assume it functions as a literal shorthand for `thing = thing + x`. Something like this would be good maybe:

    int src_address_a = dst_address_a;
    if (!is_assignment) {
        int src_address_a = _parse_expression(codegen, on->arguments[0], p_stack_level, false, p_initializer);
    }

**Comment by Calinou** [comment]
@demipixel Can you (or anyone else) still reproduce this bug in [Godot 3.2.3](https://godotengine.org/article/maintenance-release-godot-3-2-3) or any later release?

If yes, please ensure that an up-to-date Minimal Reproduction Project (MRP) is included in this report (a MRP is a zipped Godot project with the minimal elements necessary to reliably trigger the bug). You can upload ZIP files in an issue comment with a drag and drop.

**Comment by vnen** [comment]
I think I fixed something like this already.

**Comment by akien-mga** [comment]
Yes this is fixed in 3.2.3.
